Alternative to #20 (#21)
strimzi/governance: GOVERNANCE.md

Such components might be represented by a separate GitHub repository within the Strimzi organization (e.g. `strimzi-kafka-bridge`) or by a subdirectory in a GitHub repository (e.g. `./documentation/` in `strimzi-kafka-operator`).
The components owners are identified in the [`COMPONENT-OWNERS`](COMPONENT-OWNERS) file in this repository which also includes the component for which they are responsible.

### Maintainers

A maintainer is actively responsible for the overall development of the whole Strimzi project and all its components.
The maintainers are identified in the [`MAINTAINERS`](MAINTAINERS) file.

Every maintainer is, in effect, also an owner for all components and does not have to be mentioned in the `COMPONENT-OWNERS` file.

### Emeritus code owners and maintainers

An emeritus code owner, or an emeritus maintainer, is someone who is formerly had that role but is no longer actively involved with the project.
Emeritus code owners and Emeritus maintainers are identified in the [`EMERITUS`](EMERITUS) file.

In the rest of this document the term 'maintainer' should be taken to mean 'active maintainer' unless qualified using the term 'emeritus maintainer'.

> #### Changes in Maintainership

Reviewer: I don't think it makes sense to have this as a subchapter of `Emeritus code owners and maintainers`. It should be a the same level.

New maintainers are proposed by an existing maintainer and are elected by a vote (see later for the rules of this vote).
If the vote succeeds the proposed maintainer will be invited to become a maintainer.
If the proposed maintainer accepts the invitation:
* Their name will be added to the `MAINTAINERS` file.
* They will be added to the collaborators list of the Strimzi repository with "Write" access
* After 6 months they will be given "Admin" access to the Strimzi repository.

Maintainers can be removed by a vote.
If the vote succeeds:
* Their name will be removed from the `MAINTAINERS` file.
* They will be removed to the collaborators list of the Strimzi repository

Maintainers who wish to become emeritus simply mark the fact by opening a PR against this repository moving their name to the `EMERITUS` file.
The process for reverting from an emeritus maintainer to an active maintainer is the same as for new maintainers.

> #### Changes in component ownership

Reviewer: I don't think it makes sense to have this as a subchapter of `Emeritus code owners and maintainers`. It should be a the same level.

Component owners can be removed by a vote.
If the vote succeeds:
* Their name will be removed from the `COMPONENT-OWNERS` file.
* They will be removed to the collaborators list of the Strimzi repository, if any.

Component owners who wish to become emeritus simply mark the fact by opening a PR against this repository moving their name to the `EMERITUS` file.
The process for reverting from an emeritus component owners to an active component owners is the same as for new component owners.

## Voting rules

The Strimzi project employs voting to ensure no single member can dominate the project.

For formal votes, a specific statement of what is being voted on should be added to the relevant GitHub issue or PR.
Voters should indicate their yes/no vote on that issue or PR, and after a suitable period of time, the votes will be tallied and the outcome noted.

The project uses different kinds of vote for different purposes.

* PR approval votes
* Lazy majority votes
* Explicit majority votes

Who may vote, where votes are recorded and the duration of the voting also depends on the subject of vote, as explained in the following sections.

> #### PR approval rules

Reviewer: ```suggestion
### PR approval rules
```

PRs which affect only a single component can be approved by at least one positive maintainer vote and one positive vote from an owner of the given component.
If the PR author is a maintainer or an owner of given component, this counts as a vote.

### Lazy majority rules

Lazy majority votes use +1, 0, -1 votes and their fractions, where:
* +1 means "yes"
* -1 means "no"
* The numbers between +1 and -1 indicate how strongly you feel about the proposal

_(Inspired by [Apache Software Foundation voting](https://www.apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions))_

Binding votes are those cast by people in specific roles defined for that vote and count towards the approval of the subject of the vote.
Non-binding may be cast by people not in any of those roles and do not count towards approval of the subject of the vote.

The vote succeeds when there are least three +1 binding votes and no -1 binding votes.
The vote should be opened for at least 3 days to give everyone enough time to vote.

### Explicit majority rules

An explicit majority vote is simply a +1 or -1. 
The vote succeeds when ⅔ of cast binding votes are +1.

> The vote should be opened for at least 7 days to give everyone enough time to vote.

Reviewer: I'm fine with this change. But it is quite significant. So maybe you should better explain what does it change in the PR description to make sure it is obvious for everyone? Because it can get easily lost in the restructuring of the document.

Author: Done.

### Votes

#### Votes on changes to Governance

With the exception of maintainers and component owners marking themselves as emeritus, all changes in the governance repository require an **Explicit majority** of **Maintainers**.

#### Votes on adding or removing maintainers

Votes to adding or removing a maintainer require an **Explicit majority** of **Maintainers**.

#### Votes on adding or removing component owners

Votes to add or remove a component owners require an **Explicit majority** of **Maintainers**.

#### Votes on Proposals

Proposals can be opened against the [proposals repository](https://github.com/strimzi/proposals).
Proposals should cover any changes to the Strimzi project which might significantly impact its users or the project direction.

Approving a proposal requires a **Lazy majority** of **Maintainers**.
*However, non-binding votes are encouraged as a signal to maintainers of the acceptability of the proposal to the wider community.*

+1 votes can be expressed by approving the proposal PR or in the comments.

> Other votes should be expressed in the comments. For example "+1 (binding)", or for a non-maintainer "+1 (non-binding)".

Reviewer: As we talk about "other votes", it might be better to have other vote examples?

```suggestion
Other votes should be expressed in the comments. For example "-1 (binding)", or for a non-maintainer "-1 (non-binding)".
```

Author: I reworded it to make clearer that -1 votes should come with _some_ reason. 

I think giving examples could make it overly specific, for example making some reasons for a -1 somehow more accepted (by virtue of being an example) than others. But that said, did you had a specific example in mind?

## Project Roles

For governance purposes the project recognises four roles: code owners, maintainers, emeritus code owners, and emeritus maintainers.
At any one time a person may be in 0 or 1 of those roles.

### Component owners

A component owner is responsible for development of a specific subproject or component within Strimzi.
Such components might be represented by a separate GitHub repository within the Strimzi organization (e.g. `strimzi-kafka-bridge`) or by a subdirectory in a GitHub repository (e.g. `./documentation/` in `strimzi-kafka-operator`).

> The components owners are identified in the [`COMPONENT-OWNERS`](COMPONENT-OWNERS) file in this repository which also includes the component for which they are responsible.

Reviewer 2: ```suggestion
The component owners are identified in the [`COMPONENT-OWNERS`](COMPONENT-OWNERS) file in this repository which also includes the component for which they are responsible.
```

### Maintainers

A maintainer is actively responsible for the overall development of the whole Strimzi project and all its components.
The maintainers are identified in the [`MAINTAINERS`](MAINTAINERS) file.

Every maintainer is, in effect, also an owner for all components and does not have to be mentioned in the `COMPONENT-OWNERS` file.

### Emeritus code owners and maintainers

> An emeritus code owner, or an emeritus maintainer, is someone who is formerly had that role but is no longer actively involved with the project.

Reviewer 2: ```suggestion
An emeritus code owner, or an emeritus maintainer, is someone who formerly had that role but is no longer actively involved with the project.
```

> Emeritus code owners and Emeritus maintainers are identified in the [`EMERITUS`](EMERITUS) file.

Reviewer 2: ```suggestion
Emeritus code owners and emeritus maintainers are identified in the [`EMERITUS`](EMERITUS) file.
```

In the rest of this document the term 'maintainer' should be taken to mean 'active maintainer' unless qualified using the term 'emeritus maintainer'.

> ### Changes in Maintainership

Reviewer 2: ```suggestion
### Changes in maintainership
```

New maintainers are proposed by an existing maintainer and are elected by a vote (see later for the rules of this vote).
If the vote succeeds the proposed maintainer will be invited to become a maintainer.
If the proposed maintainer accepts the invitation:
* Their name will be added to the `MAINTAINERS` file.
* They will be added to the collaborators list of the Strimzi repository with "Write" access
* After 6 months they will be given "Admin" access to the Strimzi repository.

Maintainers can be removed by a vote.
If the vote succeeds:
* Their name will be removed from the `MAINTAINERS` file.
* They will be removed to the collaborators list of the Strimzi repository

> Maintainers who wish to become emeritus simply mark the fact by opening a PR against this repository moving their name to the `EMERITUS` file.

Reviewer 3: So it means that if an inactive maintainer doesn't open any PR (because he/she is just inactive, not following the project, and not having any interest to open any kind of PR on it), it will be forever maintainer even if not active?
Isn't there any "inactivity" rule to make him/her emeritus or having a PR open by another maintainer?

Author: The mechanism I'm proposing for `active -> emeritus` is entirely voluntary.

The change from ⅔ _all_ maintainer approval to ⅔ _voting_ maintainer approval means that such inactive maintainers don't prevent decisions being taken due to their inactivity (which is a problem we want to solve, even though it's not yet been a problem in practice). 

Obviously that doesn't prevent someone who hasn't been active for a long time suddenly voting on some contentious issue and swinging a decision that otherwise would have gone the other way. TBH that seems really unlikely: we certainly don't have a problem with rogue maintainers currently. And there is (and always was) a mechanism for the other maintainers to remove a maintainer. So if someone did this then the other maintainers could get rid of them anyway.

Reviewer 3: Why do you think inactivity is not a good way anymore? Also, you said "there is (and always was) a mechanism for the other maintainers to remove a maintainer." ... but it's not about making someone "emeritus", it's about removing it. Why can't we have a mechanism where a maintainer can propose another one is inactive enough to be emeritus? I think if we have just "active -> emeritus is entirely voluntary.", the inactive maintainers will be just there forever (because they are inactive and not following at all what's happening here, so even not reading this proposal and not marking themselves as emeritus) ... and it won't reflect the reality of the maintainers pool.

Reviewer: I would not have a problem with a maintainer being able to propose to move someone to emeritus. But in a situation when it is not needed (i.e. after the ⅔ of all maintainer vote is changed), I would take it as a pretty aggressive move to be honest.

(As in ... I do not have a problem to have such a rule in the governance ... but I'm not sure I would vote for it)

Reviewer 3: I was the object of such a thing on the Eclipse Hono project and I found it pretty fair. I was not involved anymore as before and the project lead sent out an email to propose me to move to be "historical committer" (as you can see here https://projects.eclipse.org/projects/iot.hono/who). I just accepted because it was reflecting the reality.

Author: > Why do you think inactivity is not a good way anymore?

It just seems unnecessary to invent a whole set of extra rules and processes. By only counting the votes of people who are actively participating (in the vote) we prevent the problem we were trying to solve, but with fewer rules.

> Why can't we have a mechanism where a maintainer can propose another one is inactive enough to be emeritus? 

Why do we need a codified mechanism to do that? We could just, you know, just ask them: "I see you're no longer active in the project, would you mind opening a PR to reflect that? Or did you intend to start contributing again soon?"

Let me flip the question around: What advantage do you see to having rules to automatically make maintainers emeritus?

Reviewer 3: I would like just to have the reality reflected in the maintainers pool.
Btw asking someone to step back could be even another way as you suggest, not sure about getting a reply btw.

The process for reverting from an emeritus maintainer to an active maintainer is the same as for new maintainers.

### Changes in component ownership

New component owners can be proposed by any maintainer and are elected by a vote.
If the vote succeeds the proposed component owner will be invited to become a component owner.
If the proposed component owner accepts the invitation:
* Their name will be added to the `COMPONENT-OWNERS` file.
* Owners of components which have their own GitHub repository will get "Write" rights for given GitHub repository and will be able to merge approved PRs.
* Owners will not get "Admin" rights on any Strimzi GitHub repositories.

Component owners can be removed by a vote.
If the vote succeeds:
* Their name will be removed from the `COMPONENT-OWNERS` file.
* They will be removed to the collaborators list of the Strimzi repository, if any.

> Component owners who wish to become emeritus simply mark the fact by opening a PR against this repository moving their name to the `EMERITUS` file.

Reviewer 2: What happens if they don't do this, but don't really contribute any more?

Author: See my response to Paolo's [question, above](https://github.com/strimzi/governance/pull/21/files#r1740154232).

> The process for reverting from an emeritus component owners to an active component owners is the same as for new component owners.

Reviewer 2: ```suggestion
The process for reverting from emeritus component owners to an active component owners is the same as for new component owners.
```

## Voting rules

The Strimzi project employs voting to ensure no single member can dominate the project.

For formal votes, a specific statement of what is being voted on should be added to the relevant GitHub issue or PR.
Voters should indicate their yes/no vote on that issue or PR, and after a suitable period of time, the votes will be tallied and the outcome noted.

The project uses different kinds of vote for different purposes.

* PR approval votes
* Lazy majority votes
* Explicit majority votes

Who may vote, where votes are recorded and the duration of the voting also depends on the subject of vote, as explained in the following sections.

### PR approval rules

PRs against most repositories may be merged after receiving at least two positive maintainer votes.

> Voting for PRs can be done using a comment in the PR or by approving (or requesting changes) the PR.

Reviewer 2: ```suggestion
Voting for PRs can be done using a comment in the PR or by approving (or requesting changes to) the PR.
```

If the PR author is a maintainer, this counts as a vote.

PRs which affect only a single component can be approved by at least one positive maintainer vote and one positive vote from an owner of the given component.
If the PR author is a maintainer or an owner of given component, this counts as a vote.

### Lazy majority rules

Lazy majority votes use +1, 0, -1 votes and their fractions, where:
* +1 means "yes"
* -1 means "no"

> * The numbers between +1 and -1 indicate how strongly you feel about the proposal

Reviewer 3: what does it mean? we can vote +0.4?

Author: Yes. This is in the current rules, so feel free to vote $+\pi^{-1}$ on this proposal if you want :stuck_out_tongue_winking_eye:

Reviewer 3: I totally forgot about this rule. I should start voting in some more eccentric way :-D

_(Inspired by [Apache Software Foundation voting](https://www.apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions))_

Binding votes are those cast by people in specific roles defined for that vote and count towards the approval of the subject of the vote.

> Non-binding may be cast by people not in any of those roles and do not count towards approval of the subject of the vote.

Reviewer 2: ```suggestion
Non-binding votes may be cast by people not in any of those roles and do not count towards approval of the subject of the vote.
```

> The vote succeeds when there are least three +1 binding votes and no -1 binding votes.

Reviewer 3: so just one -1 can screw up a majority of, for example, five +1 from other maintainers? I know it was already here for proposals but are we sure it works?

Author: > so just one -1 can screw up a majority of, for example, five +1 from other maintainers? 

Yes.

> I know it was already here for proposals but are we sure it works?

Answering very literally: We don't really have enough data to answer that meaningfully. Even if then I don't know how you'd objectively decide on what "it works" means. It clearly is a mechanism which can a plurality from getting their way. In that sense it clearly doesn't "work". But the end objective is: Does it make for better decisions than some alternative?

But I think you mean: "Are we happy with the implication that every maintainer holds a veto?" Presumably we _were_ happy with that once, because it's in our current rules. So what's changed?

More generally, I think it's easy to poke holes in any system. We're not going to get a perfect system with this revision. The important things are:

1. we have a mechanism to evolve the rules as circumstances change
2. we think this proposed change is an improvement on what went before

Reviewer 3: Shame on me but I think this was another point I forgot about the current rules (like the one I can vote 0.4 ;-)), maybe just because it never happened in the reality.
But in general, I think that in a community, it's not that great having every maintainer holding a veto. I think we still would need a majority. If just a single maintainer doesn't think the proposal is good but the rest of maintainers agree, I don't think that just that -1 should stop the overall approval. This could just get a proposal stuck, because that maintainer is asking for changes (so the -1) which others don't agree and these changes will never be applied.

Author: TBH I think I'd rather see if we can get these changes through without opening up a larger debate about the veto. That's because:

* It's not a problem we've seen in practice.
* I do think there's value in getting a consensus. The veto forces people to address the concerns of others, rather than dismissing them. 
* I think there are social dynamics which usually tend to prevent a single maintainer from using their veto unnecessarily or in a problematic way.
* If it did become a problem, there is still the option of removing a maintainer.

Feel free to open a separate discussion about it if you feel strongly about it.

Reviewer 3: > I do think there's value in getting a consensus. The veto forces people to address the concerns of others, rather than dismissing them.

Why people should address concerns of others if "others" are stuck on their positions and thanks to the veto they can stop a proposal? This is my big fear, even if as you said it has not been a problem in practice until today. But how are we going to deal with it if it happens? Removing a maintainer because he is stuck on his/her position and blocking a proposal while all other maintainers are in agreement? Tbh I don't see it that great.

The vote should be opened for at least 3 days to give everyone enough time to vote.

### Explicit majority rules

An explicit majority vote is simply a +1 or -1. 

> The vote succeeds when ⅔ of cast binding votes are +1.

Reviewer 3: Does it mean that even just only one +1 makes the vote passing if no other maintainers vote within 7 days? So we are moving from 2/3 of all maintainers to just 2/3 of voters?

Author: Great point. The current wording does have that effect.

In practice I doubt anything would pass just based off a single vote, but it is a flaw. @strimzi/maintainers does anyone object if we require the casting of at least 3 votes as a way to remedy this? I think that strikes a balance between making progress and ensuring that there's been enough consideration (and from enough viewpoints) of the subject of a vote.

Reviewer 3: Of course, +1 from me on having a minimum number of binding votes.

Reviewer: I'm fine with the minimum number of votes. But at the same time, the vote with only a single +1 vote will pass only if nobody else votes within the week. So it is not something I would worry about.

Reviewer 3: > the vote with only a single +1 vote will pass only if nobody else votes within the week

Even if we see a minimum possibility it could happen, it's anyway a door left open.

Author: I have added a requirement for a minimum of three votes.

The vote should be opened for at least 7 days to give everyone enough time to vote.

### Votes

> #### Votes on changes to Governance

Reviewer 2: ```suggestion
#### Votes on changes to governance
```

> With the exception of maintainers and component owners marking themselves as emeritus, all changes in the governance repository require an **Explicit majority** of **Maintainers**.

Reviewer 2: ```suggestion
With the exception of maintainers and component owners marking themselves as emeritus, all changes in the governance repository require an **explicit majority** of **maintainers**.
```

#### Votes on adding or removing maintainers

> Votes to adding or removing a maintainer require an **Explicit majority** of **Maintainers**.

Reviewer 2: ```suggestion
Votes to add or remove maintainers require an **explicit majority** of **maintainers**.
```

#### Votes on adding or removing component owners

> Votes to add or remove a component owners require an **Explicit majority** of **Maintainers**.

Reviewer 2: ```suggestion
Votes to add or remove component owners require an **explicit majority** of **maintainers**.
```

> #### Votes on Proposals

Reviewer 2: ```suggestion
#### Votes on proposals
```

Proposals can be opened against the [proposals repository](https://github.com/strimzi/proposals).
Proposals should cover any changes to the Strimzi project which might significantly impact its users or the project direction.

> Approving a proposal requires a **Lazy majority** of **Maintainers**.

Reviewer 2: ```suggestion
Approving a proposal requires a **lazy majority** of **maintainers**.
```

*However, non-binding votes are encouraged as a signal to maintainers of the acceptability of the proposal to the wider community.*

+1 votes can be expressed by approving the proposal PR or in the comments.
Other votes, with reasons, can be expressed in the comments. 
For example "-1 (binding) because ...", or for a non-maintainer "-1 (non-binding) because ...".

> #### Votes on PRs in Other Repos

Reviewer 2: ```suggestion
#### Votes on PRs in other repositories
```

With the exception of the governance and proposals repositories, votes on PRs in other repositories use **PR approval rules**.

> #### Votes on Other Changes

Reviewer 2: ```suggestion
#### Votes on other changes
```

> Unless specified above, all other changes to the project require an **Explicit majority** of **maintainers**.

Reviewer 2: ```suggestion
Unless specified above, all other changes to the project require an **explicit majority** of **maintainers**.
```

> Additionally, any maintainer may request that any change require an **Explicit majority** of **maintainers**

Reviewer 2: ```suggestion
Additionally, any maintainer may request that any change requires an **explicit majority** of **maintainers**.
```
